contextualized/dags/torch_notmad/torch_notmad.py

```diff
@@ -0,0 +1,286 @@
+#define Encoder
+import torch
+from torch import nn
+import pytorch_lightning as pl
+import numpy as np
+from contextualized.functions import identity_link
+torch.set_default_tensor_type(torch.FloatTensor)
+
+#local imports
+from graph_utils import project_to_dag_torch
+from torch_utils import DAG_loss, NOTEARS_loss
+
+class NGAM(nn.Module):
+    """
+    Neural Generalized Additive Model for NOTMAD
+    """
+    def __init__(self, input_dim, output_dim, width, n_hidden_layers, activation=nn.SiLU, link_fn=identity_link):
+        """ Initialize the NGAM encoder.
+
+        Args:
+            input_dim (int, 1): Input shape of Encoder
+            output_dim (int, ): Output shape of Encoder (weights for Explainer)
+            width (int): Size of hidden layers
+            n_hidden_layers (int): # of hidden layers
+            activation (nn.Module: nn.SiLU): Activation function to use.
+            link_fn (lambda: identity_link): Link function to use.
+        """
+        super(NGAM, self).__init__()
+        self.input_dim = input_dim
+        self.output_dim = output_dim
+        
+        nam_input_layer = nn.Linear(1, width)
+        init_hidden_layer = nn.Linear(width, width)
+        nam_output_layer = nn.Linear(width, output_dim)
+        
+        with torch.no_grad(): #use uniform randomization
+            init_hidden_layer.weight = self._get_new_weights(init_hidden_layer)
+            nam_input_layer.weight = self._get_new_weights(nam_input_layer)
+            nam_output_layer.weight = self._get_new_weights(nam_output_layer)
+        
+        hidden_layers = lambda: [layer for layer in (init_hidden_layer, activation()) for _ in range(n_hidden_layers)]
+        nam_layers = lambda: [nam_input_layer, activation()] + hidden_layers() + [nam_output_layer]
+        
+        self.nams =  nn.ModuleList([nn.Sequential(*nam_layers()) for _ in range(self.input_dim)])
+        self.link_fn = link_fn
+
+    def forward(self, x):
+        batch_size = x.shape[0]
+        ret = torch.zeros((batch_size, self.output_dim))
+        for i, nam in enumerate(self.nams):
+            ret += nam(x[:, i].unsqueeze(-1)) #batch at contextual feature i
+        
+        return self.link_fn(ret)
+    
+    def _get_new_weights(self, layer_):
+        with torch.no_grad():
+            return nn.parameter.Parameter(
+                    torch.tensor(np.random.uniform(-0.01, 0.01, size=layer_.weight.shape)).float()
+                )
+    
+
+class Explainer(nn.Module):
```

Comment:
Remove and use the modules.py Explainer. May need to add some extra functionality to the Explainer to get/set the archetypes.

Reply:
I just updated the explained to include archetype getters and setters:
https://github.com/cnellington/Contextualized/blob/e5fe0a07fe18b6d4df843dc01820cb3b3087b494/contextualized/modules.py#L39

Last missing piece is the mask used by NOTMAD.  I'd suggest either including the mask in NOTMAD or creating a subclass of the modules.py Explainer that extends the functionality.